jerry-main/main-unix-snapshot.c

```diff
@@ -0,0 +1,413 @@
+        }
+      }
+    }
+
+    buffer_p += code_size;
+  }
+  while (buffer_p < buffer_end_p);
+} /* scan_snapshot_functions */
+
+/**
+ * Update all literal offsets in a snapshot data.
+ */
+static void
+update_literal_offsets (uint8_t *buffer_p, /**< snapshot buffer start */
+                        const uint8_t *buffer_end_p, /**< snapshot buffer start */
+                        lit_mem_to_snapshot_id_map_entry_t *lit_map_p) /**< literal map */
+{
+  JERRY_ASSERT (buffer_end_p > buffer_p);
+
+  do
+  {
+    ecma_compiled_code_t *bytecode_p = (ecma_compiled_code_t *) buffer_p;
+    uint32_t code_size = ((uint32_t) bytecode_p->size) << JMEM_ALIGNMENT_LOG;
+
+    if (bytecode_p->status_flags & CBC_CODE_FLAGS_FUNCTION)
+    {
+      jmem_cpointer_t *literal_start_p;
+      uint32_t const_literal_end;
+
+      if (bytecode_p->status_flags & CBC_CODE_FLAGS_UINT16_ARGUMENTS)
+      {
+        literal_start_p = (jmem_cpointer_t *) (buffer_p + sizeof (cbc_uint16_arguments_t));
+
+        cbc_uint16_arguments_t *args_p = (cbc_uint16_arguments_t *) buffer_p;
+        const_literal_end = args_p->const_literal_end;
+      }
+      else
+      {
+        literal_start_p = (jmem_cpointer_t *) (buffer_p + sizeof (cbc_uint8_arguments_t));
+
+        cbc_uint8_arguments_t *args_p = (cbc_uint8_arguments_t *) buffer_p;
+        const_literal_end = args_p->const_literal_end;
+      }
+
+      for (uint32_t i = 0; i < const_literal_end; i++)
+      {
+        if (literal_start_p[i] != JMEM_CP_NULL)
+        {
+          lit_mem_to_snapshot_id_map_entry_t *current_p = lit_map_p;
+
+          while (current_p->literal_id != literal_start_p[i])
+          {
+            current_p++;
+          }
+
+          literal_start_p[i] = current_p->literal_offset;
+        }
+      }
+    }
+
+    buffer_p += code_size;
+  }
+  while (buffer_p < buffer_end_p);
+} /* update_literal_offsets */
+
+/**
+ * Print help.
+ */
+static void
+print_help (char *name) /**< program name */
+{
+  printf ("Usage: %s [COMMAND] [ARGUMENTS]\n"
+          "\n"
+          "Command list:\n"
+          "  merge [OUTPUT_FILE] [INPUT_FILEs]...\n"
+          "      Merge input files into output file.\n"
+          "  help\n"
+          "      Prints this help.\n"
+          "\n",
+          name);
+} /* print_help */
+
+/**
+ * Main function.
+ *
+ * @return error code (0 - no error)
+ */
+int
+main (int argc, /**< number of arguments */
+      char **argv) /**< argument list */
+{
+  if (argc <= 1 || (argc == 2 && (!strcmp ("help", argv[1]))))
+  {
+    print_help (argv[0]);
+    return JERRY_STANDALONE_EXIT_CODE_OK;
+  }
+
+  if (strcmp ("merge", argv[1]))
+  {
+    printf ("Unknown command: %s\n", argv[1]);
+    print_help (argv[0]);
+    return JERRY_STANDALONE_EXIT_CODE_OK;
+  }
+
+  if (argc < 5)
+  {
+    printf ("At least an output and two input files must be passed\n");
+    return JERRY_STANDALONE_EXIT_CODE_FAIL;
+  }
```

Comment:
We still disagree. Sub-commands in programs are intuitive only if they combine various really different functionalities in a single tool. Like git: `git branch` has semantics and CLI syntax very different from `git add` or `git clone`. (git pushes that to the extremes, though, having sub-sub-commands at some places, like `git submodule update`.) But then, even in git there is the option syntax after the sub-(sub-)commands.

And in this tool, there is only one functionality: merging snapshots. Printing help does not count as another, IMO.

(Even if there would be more sub-commands in this tool, it would probably be highly beneficial to have proper CLI for each, which is possible with cli.c by starting argv-processing from the 2nd index, roughly by `cli_init (snap_opts, argc - 2, argv + 2);`)

As for the merging functionality, a linker tool is a much better analogy. There, just like here, you have (potentially multiple) inputs and a single output. And the usual syntax is `tool -o outfile infile1 infile2...` _or_ `tool infile1 infile2... -o outfile` (or, it is possible to put the `-o outfile` part anywhere, actually). I'd argue that for those tools `-o` is mandatory (omitting it and producing a.out should not really be a valid option). But having it as an option allows better flexibility and user-error-proofness on the command line.

Finally, two args with an option go with:
```c
  CLI_OPT_DEF (.id = OPT_XXX, .longopt = '--xxx', .argc = 2, .meta = "META1 META2",
               .help = "xxx"),
```
and they will be available as `cli_state.arg[1]` and `cli_state.arg[2]`